fdbserver/BlobManifest.actor.cpp

```diff
@@ -0,0 +1,369 @@
+
+#define ENABLE_DEBUG_PRINT true
+template <typename... T>
+inline void dprint(fmt::format_string<T...> fmt, T&&... args) {
+	if (ENABLE_DEBUG_PRINT)
+		fmt::print(fmt, std::forward<T>(args)...);
+}
+
+// This class dumps blob manifest to external blob storage.
+class BlobManifestDumper : public ReferenceCounted<BlobManifestDumper> {
+public:
+	BlobManifestDumper(Database& db, Reference<BlobConnectionProvider> blobConn) : db_(db), blobConn_(blobConn) {}
+	virtual ~BlobManifestDumper() {}
+
+	// Execute the dumper
+	ACTOR static Future<Void> execute(Reference<BlobManifestDumper> self) {
+		try {
+			state Standalone<BlobManifest> manifest;
+			Standalone<VectorRef<KeyValueRef>> rows = wait(getSystemKeys(self));
+			manifest.rows = rows;
+			Value data = encode(manifest);
+			wait(writeToFile(self, data));
+		} catch (Error& e) {
+			dprint("WARNING: unexpected blob manifest dumper error {}\n", e.what()); // skip error handling for now
+		}
+		return Void();
+	}
+
+private:
+	// Return system keys that to be backed up
+	ACTOR static Future<Standalone<VectorRef<KeyValueRef>>> getSystemKeys(Reference<BlobManifestDumper> self) {
+		state Standalone<VectorRef<KeyValueRef>> rows;
+		loop {
+			state Transaction tr(self->db_);
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			try {
+				state std::vector<KeyRangeRef> ranges = {
+					blobGranuleMappingKeys, // Map granule to workers. Track the active granules
+					blobGranuleFileKeys, // Map a granule version to granule files. Track files for a granule
+					blobGranuleHistoryKeys, // Map granule to its parents and parent bundaries. for time-travel read
+					blobRangeKeys // Key ranges managed by blob
+				};
+				for (auto range : ranges) {
+					RangeResult result = wait(tr.getRange(range, GetRangeLimits::BYTE_LIMIT_UNLIMITED));
+					for (auto& row : result) {
+						rows.push_back_deep(rows.arena(), KeyValueRef(row.key, row.value));
+					}
+				}
+				return rows;
+			} catch (Error& e) {
+				wait(tr.onError(e));
+			}
+		}
+	}
+
+	// Write data to blob manifest file
+	ACTOR static Future<Void> writeToFile(Reference<BlobManifestDumper> self, Value data) {
+		state Reference<BackupContainerFileSystem> writer;
+		state std::string fileName;
+
+		std::tie(writer, fileName) = self->blobConn_->createForWrite(MANIFEST_FILENAME);
+		state Reference<IBackupFile> file = wait(writer->writeFile(fileName));
+		wait(file->append(data.begin(), data.size()));
+		wait(file->finish());
+		dprint("Write blob manifest file with {} bytes\n", data.size());
+		return Void();
+	}
+
+	// Encode manifest as binary data
+	static Value encode(BlobManifest& manifest) {
+		BinaryWriter wr(IncludeVersion(ProtocolVersion::withBlobGranuleFile()));
+		wr << manifest;
+		return wr.toValue();
+	}
+
+	Database db_;
+	Reference<BlobConnectionProvider> blobConn_;
+};
+
+// Defines granule info that interests full restore
+struct BlobGranuleVerion {
```

Comment:
Nit: Typo> This should be versoin instead of verion?

Reply:
fixed. thanks